File: Data Structures & Algorithms/search-2d-matrix/submission-4.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();
        int n = matrix[0].size();

        int top = 0, bot = m-1;
        int row;
        while (top <= bot) {
            row = (top+bot)/2;

            if (target < matrix[row][0]) {
                bot = row - 1;
                continue;
            } else if (target <= matrix[row][n-1]) {
                break;
            } else top = row + 1;
        }

        int left = 0, right = n-1;
        int mid;
        while (left<=right) {
            mid = (left+right)/2;

            if (target < matrix[row][mid]) {
                right = mid -1;
                continue;
            } else if(matrix[row][mid] == target) {
                return true;
            } else left = mid + 1;
        }
        return false;
    }
};

File: Data Structures & Algorithms/spiral-matrix/submission-4.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m = len(matrix)
        n = len(matrix[0])

        ref = 0
        ans = []

        minn = min(m, n)

        while ref < (minn+1)//2:
            #top
            for i in range(ref, n-ref):
                ans.append(matrix[ref][i])

            #right
            for i in range(ref+1, m-ref):
                ans.append(matrix[i][n-ref-1])

            #bottom
            if ref < minn//2:
                for i in range(n-ref-2, ref-1, -1):
                    ans.append(matrix[m-ref-1][i])

                #left
                for i in range(m-ref-2, ref, -1):
                    ans.append(matrix[i][ref])

            ref += 1

        return ans
        